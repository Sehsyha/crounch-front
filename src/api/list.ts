import { api } from './api';

export function createList(name: string) {
  return api
    .post('lists', { name })
    .then(res => res.data);
}
