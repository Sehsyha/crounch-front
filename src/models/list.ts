import { Product } from './product';
import { CategoryInGetListResponse } from './category';

interface List {
  id: string
  name: string
  products: Product[]
}

interface GetListResponse {
  id: string
  name: string
  creationDate: string
  categories: CategoryInGetListResponse[]
}

export { List, GetListResponse }
