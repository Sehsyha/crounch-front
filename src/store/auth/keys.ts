export enum AuthActions {
  LOGIN = 'LOGIN',
  SIGNUP = 'SIGNUP',
  LOGOUT = 'LOGOUT',
}

export enum AuthGetters {
  IS_AUTHENTICATED = 'IS_AUTHENTICATED',

}

export enum AuthMutations {
  LOGIN = 'LOGIN',
  LOGOUT = 'LOGOUT',
}
