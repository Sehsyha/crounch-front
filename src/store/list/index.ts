import { Module } from 'vuex';

import { actions } from './actions';
import { RootState } from '..';
import { List } from '@/models/list';

export interface ListState {
  lists: List[];
}

export const state: ListState = {
  lists: [],
};

const namespaced: boolean = true;
export const listNamespace = { namespace: 'list' };

export const list: Module<ListState, RootState> = {
  namespaced,
  state,
  actions,
};
