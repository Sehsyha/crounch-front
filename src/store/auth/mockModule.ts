import { AuthKeys } from './keys';
import { Module } from 'vuex';
import { AuthState } from '.';
import { RootState } from '..';

const actions = {
  [AuthKeys.SIGNUP]: jest.fn(),
  [AuthKeys.LOGIN]: jest.fn(),
  [AuthKeys.LOGOUT]: jest.fn(),
};

const getters = {
  isAuthenticated: jest.fn(),
};

export function createAuthModuleMock(): Module<AuthState, RootState> {
  return {
    namespaced: true,
    actions,
    getters,
    state: {
      token: '',
      status: '',
    },
  };
}
