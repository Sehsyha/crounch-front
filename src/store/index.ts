import Vue from 'vue';
import Vuex, { StoreOptions } from 'vuex';
import { auth } from './auth';
import { ListModule } from './ListModule';
import { product } from './product'

Vue.use(Vuex);

export interface RootState { }

const store: StoreOptions<RootState> = {
  state: {},
  modules: {
    auth,
    ListModule,
    product,
  },
  strict: true,
};

export default new Vuex.Store<RootState>(store);
