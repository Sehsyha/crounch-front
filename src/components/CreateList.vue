<template>
  <div id="create-list">
    <h2>Création de liste</h2>
    <div v-if="!isNameValid" id="name-error">Le nom doit être renseigné</div>
    <div v-if="result !== ''" class="result">{{ result }}</div>
    <form @submit.prevent="createList">
      <input type="text" placeholder="Nom" v-model="name" />
      <input type="submit" value="Créer" />
    </form>
  </div>
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator';
import { getModule } from 'vuex-module-decorators';

import { ListModule } from '@/store/ListModule';

@Component
export default class CreateList extends Vue {
  public listModule: ListModule = getModule(ListModule)
  name: string = '';
  result: string = '';

  async createList() {
    if (!this.isNameValid) {
      return;
    }

    await this.listModule.create(this.name)
    this.result = 'Créée'
    this.name = ''
  }

  get isNameValid(): boolean {
    return this.name.length > 0;
  }
}
</script>
