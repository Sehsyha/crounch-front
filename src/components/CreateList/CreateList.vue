<template>
  <div id="create-list">
    <div v-if="!isNameValid" id="name-error">Le nom doit être renseigné</div>
    <div v-if="result !== ''" class="result">{{ result }}</div>
    <form @submit.prevent="createList">
      <input type="text" placeholder="Nom" v-model="name" />
      <input type="submit" value="Créer" />
    </form>
  </div>
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator';
import { Action } from 'vuex-class';

import { ListKeys } from '@/store/list/keys';
import { listNamespace } from '@/store/list';

@Component
export default class CreateList extends Vue {
  @Action(ListKeys.CREATE, listNamespace) doCreateList: any;
  name: string = '';
  result: string = '';

  createList() {
    if (!this.isNameValid) {
      return;
    }

    this.doCreateList({ name: this.name }).then(() => {
      this.result = 'Créée';
    });
  }

  get isNameValid(): boolean {
    return this.name.length > 0;
  }
}
</script>
