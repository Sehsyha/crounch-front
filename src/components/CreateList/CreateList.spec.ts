import { Wrapper } from '@vue/test-utils';
import { Module } from 'vuex';

import { shallowComponent } from '@/utils/test';
import CreateList from './CreateList.vue';
import { ListKeys } from '@/store/list/keys';
import { createListModuleMock } from '@/store/list/mockModule';
import { ListState } from '@/store/list';
import { RootState } from '@/store';

jest.mock('@/store/list');

describe('CreateList', () => {
  let wrapper: Wrapper<CreateList>;
  let list: Module<ListState, RootState>;
  const name = 'shopping list';

  beforeEach(() => {
    list = createListModuleMock();

    const modules = {
      list,
    };

    wrapper = shallowComponent(CreateList, { modules });
  });

  afterEach(() => {
    wrapper.destroy();
  });

  it('Should render.', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('State', () => {
    it('Should initialize list name state to empty string.', () => {
      expect(wrapper.vm.$data.name).toBe('');
    });

    it('Should initialize result state to empty string.', () => {
      expect(wrapper.vm.$data.result).toBe('');
    });
  });

  describe('Form validation', () => {
    it('Should render an error when name is not valid.', () => {
      expect(wrapper.find('#name-error').exists()).toBeTruthy();
    });

    it('Should not render an error when name is valid.', () => {
      wrapper.setData({ name });

      expect(wrapper.find('#name-error').exists()).toBeFalsy();
    });
  });

  it('Should not dispatch create list store action when form is not valid.', () => {
    wrapper.find('[type=submit]').trigger('click');

    expect(list.actions![ListKeys.CREATE] as jest.Mock).not.toHaveBeenCalled();
  });

  it('Should dispatch create list store action when form is valid and submitted.', done => {
    wrapper.setData({ name });

    wrapper.find('[type=submit]').trigger('click');

    setTimeout(() => {
      expect(list.actions![ListKeys.CREATE] as jest.Mock).toHaveBeenCalledWith(
        expect.anything(),
        { name },
      );
      done();
    });
  });

  it('Should display result message when request succeed.', done => {
    wrapper.setData({ name });

    wrapper.find('[type=submit]').trigger('click');

    setTimeout(() => {
      expect(wrapper.find('.result').text()).toEqual('Créée')
      done();
    });
  });

  it('Should reset name when request succeed.', done => {
    wrapper.setData({ name });

    wrapper.find('[type=submit]').trigger('click');

    setTimeout(() => {
      expect(wrapper.vm.$data.name).toBe('');
      done();
    });
  });
});
