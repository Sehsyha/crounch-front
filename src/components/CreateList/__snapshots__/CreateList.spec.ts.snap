// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateList Should render. 1`] = `
<div id="create-list">
  <div id="name-error">Le nom doit être renseigné</div>
  <!---->
  <form><input type="text" placeholder="Nom"> <input type="submit" value="Créer"></form>
</div>
`;
